Change Y-axis to scientific notation if numbers are too small or too …

javerbukh opened the pull request:
…large

<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our code of conduct,
https://github.com/spacetelescope/jdaviz/blob/main/CODE_OF_CONDUCT.md . -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable viz component(s). -->

This pull request is to address numbers not showing correctly in the Y-axis of jdaviz's profile viewers.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

* Fixes #328
* Fixes #142
* Fixes #848 

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [ ] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [ ] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [ ] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

pllim commented:
Does not seem to fix anything for me in the Mosviz example notebook:

![Screenshot 2021-09-14 143539](https://user-images.githubusercontent.com/2090236/133314540-a36aeecb-f5bf-44b7-9c2d-316ae5bd8154.jpg)

pllim commented:
Nor Specviz when I zoom it out a lot:

![Screenshot 2021-09-14 143752](https://user-images.githubusercontent.com/2090236/133314801-b5b361ac-ba90-4098-88b5-dfbdfe3c408d.jpg)

javerbukh commented:
@pllim I noticed with the Mosviz example notebook the code to highlight the first row does not run the same code as actually clicking the row. Because of this, the code that fixes the scientific notation in that axes does not get called. I'm not sure if improving that code is part of this PR or not. I'll check out the specviz example. Which data are you using for that?

javerbukh commented:
Never mind it clicked...is there a reason the user would zoom out to the point that their spectrum is flat?

pllim commented:
> improving that code is part of this PR or not

Well, we can always open follow-up issue, but I think a lot of users will just see what is first presented and then notice the bad axis labels and write us help calls.

> is there a reason the user would zoom out to the point that their spectrum is flat?

No, but I do not assume that user behaviors make sense. But more seriously, axes behavior should not matter if the data makes sense or not, does that make sense?

ibusko reviewed (commented):
I couldn't load the mosviz example notebook with this branch. I tried both the nirspec and the niriss data sets, with and without images, with remote data or the same data pulled locally. I always get a stack trace that ends with this wcs-related error:

`~/Projects/software/miniconda3_2/envs/jdaviz3.8/lib/python3.8/site-packages/glue/core/coordinate_helpers.py in axis_label(wcs, axis)
    164 def axis_label(wcs, axis):
    165 
--> 166     if wcs.world_axis_names[axis] != '':
    167         return wcs.world_axis_names[wcs.world_n_dim - 1 - axis].title()
    168 

IndexError: list index out of range`

javerbukh commented:
@ibusko That traceback looks unrelated to this PR. Did you `pip install` before testing?

ibusko commented:
@javerbukh right.....I forgot to install. Now it works, and the 2d spectrum plot looks awesome. Is this being applied to the 1d spectrum as well? It looks a bit crowded still:

<img width="613" alt="image" src="https://user-images.githubusercontent.com/3577145/133453154-51da0188-f613-4c6f-9efd-4f905bda22c3.png">

pllim commented:
There is a conflict, needs a rebase.

Also, does using glue-viz/glue-jupyter#247 change anything here?

javerbukh commented:
@pllim I left a message in that PR that it actually prevents the Y-axis from updating in Mosviz at all. I'll rebase and change a couple things.

javerbukh commented:
Fixed the conflict and squashed commits. I also made scientific notation the default for the Y-axis in specviz and left a note on how to disable it in the `SpecvizExample` notebook. I am still seeing the Mosviz spectrum viewer start off as regular notation until the state changes in some way (pan, Y-axis zoom, row change, etc.) but that might be a bug in bqplot. Pinging @mariobuikhuizen and @maartenbreddels to see if they know why this is happening.

ibusko commented:
I like the default for the Y-axis in specviz to be in scientific notation. In many cases this axis is expressed in physical units that cause the numbers to be very small. At least in the UV-optical, it's usually flam or fnu.  I wonder how much variability there is in the units generated by the JWST pipelines. If micro-Jy, numbers are ~1.

javerbukh commented:
@ibusko That's a good point, I'm not as familiar with all of the use cases so I wasn't sure which to make the default (scientific notation or no). I will say that a future PR could put a toggle for that option somewhere in the UI (maybe those 3 dots finally get used for something), but I do not think I have the time to figure out how to do that by tomorrow COB.

pllim commented on jdaviz/configs/specviz/helper.py:
```diff
@@ -229,6 +229,28 @@ def _redshift_listener(self, msg):
         if msg.param == "redshift":
             self._redshift = msg.value
 
+    def activate_scientific_notation(self, activate=True, axis=None):
```
This still hardcodes the formatting, in a sense. I wonder if this can be generalized. 
```suggestion
    def set_spectrum_tick_format(self, fmt, axis=None):
```

pllim commented on jdaviz/configs/specviz/helper.py:
```diff
@@ -229,6 +229,28 @@ def _redshift_listener(self, msg):
+            logging.warning("Please use either 0 or 1 for the axis value")
+            return
+
+        if activate:
+            self.app.get_viewer("spectrum-viewer").figure.axes[axis].tick_format = '0.1e'
+        else:
+            self.app.get_viewer("spectrum-viewer").figure.axes[axis].tick_format = '0.2f'
```
This will allow user to set whatever they want, without us having to assume anything.
```suggestion
        self.app.get_viewer("spectrum-viewer").figure.axes[axis].tick_format = fmt
```

javerbukh replied:
Ok, I switched it to this implementation.

pllim reviewed (commented):
Crazy idea below.

pllim commented on jdaviz/configs/specviz/helper.py:
```diff
@@ -229,6 +229,25 @@ def _redshift_listener(self, msg):
+        ----------
+        fmt : str
+            Format of tick marks in the spectrum viewer.
+        axis
```
```suggestion
        axis : {0, 1}
```

pllim commented on jdaviz/configs/specviz/helper.py:
```diff
@@ -229,6 +229,25 @@ def _redshift_listener(self, msg):
+        fmt : str
+            Format of tick marks in the spectrum viewer.
+        axis
+            The Specutils data axis.
```
"specutils" does not quite make sense to me when I read it. I suggested something but if that is inaccurate, feel free to correct it.
```suggestion
            The spectrum viewer data axis.
            Axis 1 corresponds to the Y-axis and 0 to the X-axis.
```

pllim commented on jdaviz/configs/specviz/helper.py:
```diff
@@ -229,6 +229,25 @@ def _redshift_listener(self, msg):
+        Parameters
+        ----------
+        fmt : str
+            Format of tick marks in the spectrum viewer.
```
```suggestion
            Format of tick marks in the spectrum viewer.
            For example, ``'0.1e'`` to set scientific notation or ``'0.2f'`` to turn it off.
```

pllim reviewed (commented):
Minor comments.

javerbukh commented:
@pllim Implemented the suggestions, thank you!

pllim reviewed (approved):
I think it makes sense to use scientific notation for science. 😸

rosteen commented:
I'm still perplexed by the axis labels not being in the right format initially in Mosviz, but I'm willing to approve this anyway as an improvement on the current state. @javerbukh it's probably ok to have an example of this in the example notebook for now, but would you create a follow-up issue to add this to the docs proper?

javerbukh commented:
@rosteen Will do. My theory is that highlighting a row does not run all of the same methods as actually clicking it, but looking at the code it should "just work". I am equally perplexed.
